From 0ef729ec1b6e84d79c1eed0aa7d77a4842181c1b Mon Sep 17 00:00:00 2001
From: Elvis Lee <kwangwoong.lee@lge.com>
Date: Wed, 10 Dec 2014 15:28:17 +0900
Subject: [PATCH] Prevent items from being deleted while removing

Delegate items can be deleted when remove them from cache list.
That may cause to crash. ObjectReference keep the items from being deleted.
---
 src/qml/types/qqmldelegatemodel.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/qml/types/qqmldelegatemodel.cpp b/src/qml/types/qqmldelegatemodel.cpp
index 7a12813f0..14136d5d0 100644
--- a/src/qml/types/qqmldelegatemodel.cpp
+++ b/src/qml/types/qqmldelegatemodel.cpp
@@ -1401,6 +1401,12 @@ void QQmlDelegateModel::_q_itemsRemoved(int index, int count)
 
     d->m_count -= count;
     const QList<QQmlDelegateModelItem *> cache = d->m_cache;
+    //Prevents items being deleted in remove loop
+    for (int i = 0; i < cache.count(); ++i) {
+        QQmlDelegateModelItem *item = cache.at(i);
+        item->referenceObject();
+    }
+
     for (int i = 0, c = cache.count();  i < c; ++i) {
         QQmlDelegateModelItem *item = cache.at(i);
         // layout change triggered by removal of a previous item might have
@@ -1413,6 +1419,11 @@ void QQmlDelegateModel::_q_itemsRemoved(int index, int count)
         else  if (item->modelIndex() >= index)
             item->setModelIndex(-1);
     }
+    //Release items which are referenced before the loop
+    for (int i = 0; i < cache.count(); ++i) {
+        QQmlDelegateModelItem *item = cache.at(i);
+        item->releaseObject();
+    }
 
     QVector<Compositor::Remove> removes;
     d->m_compositor.listItemsRemoved(&d->m_adaptorModel, index, count, &removes);
